src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs

```diff
@@ -245,9 +245,141 @@ void addArg(RefKind refKind, BoundExpression expression)
+                                ImmutableArray.Create(input, makeExplicitRangeArgument(e)));
+
+                            var outputTemp = new BoundDagTemp(e.Syntax, e.SliceType, e);
+                            BoundExpression output = _tempAllocator.GetTemp(outputTemp);
+                            return _factory.AssignmentExpression(output, callExpr);
+                        }
+
+                    case BoundDagSliceEvaluation { IndexerInfo: { Symbol: MethodSymbol sliceMethod } } e:
+                        {
+                            // implicit Range indexer access via Slice method
+
+                            Debug.Assert(sliceMethod.ContainingSymbol.Equals(input.Type));
+                            Debug.Assert(sliceMethod.ParameterCount == 2);
+                            Debug.Assert(sliceMethod.Parameters[0].Type.SpecialType == SpecialType.System_Int32);
+                            Debug.Assert(sliceMethod.Parameters[1].Type.SpecialType == SpecialType.System_Int32);
+                            Debug.Assert(e.StartIndex >= 0 && e.EndIndex <= 0);
+
+                            var lengthArg = _factory.IntSubtract(_tempAllocator.GetTemp(e.LengthTemp), _factory.Literal(e.StartIndex - e.EndIndex));
+                            var callExpr = _factory.Call(input, sliceMethod, _factory.Literal(e.StartIndex), lengthArg);
+
+                            var outputTemp = new BoundDagTemp(e.Syntax, e.SliceType, e);
+                            BoundExpression output = _tempAllocator.GetTemp(outputTemp);
+                            return _factory.AssignmentExpression(output, callExpr);
+                        }
+
+                    case BoundDagSliceEvaluation { IndexerInfo: { Symbol: PropertySymbol indexer } info } e:
+                        {
+                            // explicit Range indexer access via applicable indexer member
+
+                            var callExpr = makeIndexerCall(e.Syntax, info, indexer, makeExplicitRangeArgument(e));
+                            var outputTemp = new BoundDagTemp(e.Syntax, e.SliceType, e);
+                            BoundExpression output = _tempAllocator.GetTemp(outputTemp);
+                            return _factory.AssignmentExpression(output, callExpr);
+                        }
+
                     default:
                         throw ExceptionUtilities.UnexpectedValue(evaluation);
                 }
+
+                BoundExpression makeImplicitIndexArgument(int index, BoundDagTemp lengthTemp)
+                {
+                    return index < 0
+                        ? _factory.IntSubtract(_tempAllocator.GetTemp(lengthTemp), _factory.Literal(-index))
+                        : _factory.Literal(index);
+                }
+
+                BoundExpression makeExplicitIndexArgument(int index, bool fromEnd)
+                {
+                    return _factory.New(
+                        WellKnownMember.System_Index__ctor,
```

Comment:
Have we already checked that this well-known member is present?
Same question for `Range` below (line 347).